### DISCUSSION TITLE
Pulse counting question

### DISCUSSION THREAD

**Discussion by ethos-jgreenwood** [Moddable-OpenSource/moddable#1374] [Q&A] [answered]
I see there is a PulseCount class and that it is implemented for nrf52. It appears that this is setup for quadrature decoding and not just general counting. 

Does anyone know a way to use the PulseCount class for single pin counting? Setting control to null slows the system down drastically, but leaving it off throws an error. Setting the signal and control pins to the same pin doesn't appear to count anything.

To be clear, I know the hardware is working because I've setup a Digital pin with onReadable counting -- this works except if I ever do Timer.delay -- then counts are missed (obviously).

**Comment by phoddie** [comment]
> Setting control to null slows the system down drastically, but leaving it off throws an error. 

The `control` pin is [required](https://419.ecma-international.org/#-10-io-classes-pulse-count). The implementation coerces the value to a number. Passing `null` ends up being equivalent to passing `0` (since JavaScript coercion rules make `null` become zero).

> It appears that this is setup for quadrature decoding and not just general counting.

That's correct. There's no intention for the `PulseCount` API to be used as a single pin counter.

It doesn't appear that the [`nrf_qdec` driver](https://developer.nordicsemi.com/nRF_Connect_SDK/doc-legacy/2.6.99-cs2/nrfx/drivers/qdec/driver.html) supports that behavior either.

The ECMA-419 Digital input doesn't provide a counter. But... the older Digital Monitor class on nRF52 does implement a count of rises and falls (based on interrupt callbacks). We don't generally recommend using the original Moddable SDK IO classes, but here it might be useful.

**Reply by ethos-jgreenwood** [reply]
The Digital Monitor seems to have exactly what I want.

I implemented it as follows:
```js
    this.#pinMonitor = new Monitor({
      pin: sensor_pin,
      mode: Digital.Input,
      edge: Monitor.Falling,
      onChanged() {
        trace(`Pin monitor: ${this.falls}`);
      }
    });
```

Then periodically:
```js
let count = this.#pinMonitor.falls;
trace(`Count: ${count}`);
```

This seems to work fine, I get "Count: 0". However, when I send a pulse, I get "# Break: (host): no function!". I made sure to put 		"$(MODULES)/pins/digital/monitor/manifest.json", in my host's manifest.json under "include". If I don't send a pulse, I don't get that message ever.

**Reply by phoddie** [reply]
> The Digital Monitor seems to have exactly what I want.
>I implemented it as follows:

Glad it does what you need. The implementation looks reasonable.

> I get "Count: 0". However, when I send a pulse, I get "# Break: (host): no function!".

I'm not sure what you mean by "send a pulse".  Would you share the code?

That message indicates that you are trying to call a function that doesn't exist. When it breaks in xsbug, it will show you the line that is causing the exception.

**Reply by ethos-jgreenwood** [reply]
By "send a pulse" I mean driving the sensor_pin low externally for a short amount of time and then holding the pin high otherwise.

I simplified my code so that nothing else is running. No pulse, no error. When a pulse is sent, immediate break. Does not show the line that is breaking (see screenshot). This is on the mcrun branch and linux if that helps.

![image](https://github.com/user-attachments/assets/f128b7c6-ab86-4541-abde-b6fc954280b0)

(Edit: accidentally put this as an answer; fixed)

**Reply by phoddie** [reply, marked as answer]
Thank you for the additional details.

The monitor class expects to be used with an `onChanged` callback to notify the client when the counters change. The exception happens when the monitor attempts to report the change but there is no `onChanged` callback. Your code **tries** to do this, but uses the ECMA-419 API stye which isn't implemented by the older Monitor API. Something like this should work as you intend:

```js
    this.#pinMonitor = new Monitor({
      pin: sensor_pin,
      mode: Digital.Input,
      edge: Monitor.Falling
    });
    this.#pinsMonitor.onChanged = function() {
        trace(`Pin monitor: ${this.falls}`);
    };
```

The [documentation](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/pins/pins.md#monitor) for the `Monitor` class describes this correctly.
